lib/Infer/DataflowPruning.cpp

```diff
@@ -0,0 +1,335 @@
+    case souper::Inst::Eq:
+      return {{1, args[0].getValue() == args[1].getValue()}};
+    case souper::Inst::Ne:
+      return {{1, args[0].getValue() != args[1].getValue()}};
+    case souper::Inst::Ult:
+      return {{1, args[0].getValue().ult(args[1].getValue())}};
+    case souper::Inst::Slt:
+      return {{1, args[0].getValue().slt(args[1].getValue())}};
+    case souper::Inst::Ule:
+      return {{1, args[0].getValue().ule(args[1].getValue())}};
+    case souper::Inst::Sle:
+      return {{1, args[0].getValue().sle(args[1].getValue())}};
+    case souper::Inst::CtPop:
+      return {llvm::APInt(Inst->Width, args[0].getValue().countPopulation())};
+    case souper::Inst::Ctlz:
+      return {llvm::APInt(Inst->Width,
+                          args[0].getValue().countLeadingZeros())};
+    case souper::Inst::Cttz:
+      return {llvm::APInt(Inst->Width,
+                          args[0].getValue().countTrailingZeros())};
+    case souper::Inst::BSwap:
+      return {args[0].getValue().byteSwap()};
+    default:
+      return EvalValue(); // Indicates an 'unavailable' value
+  }
+}
+}
+
+// Errs on the side of an 'unavailable' result
+// TODO: Some instructions unimplemented
+EvalValue souper::dataflow::evaluateInst(souper::Inst* Root,
+                                         ValueCache &Cache) {
+  std::vector<EvalValue> EvaluatedArgs;
+  EvalValue Result;
+
+  if (Root->K == souper::Inst::Var) {
+      Result = Cache[Root];
+  } else {
+    for (auto &&I : Root->Ops) {
+      auto eval = evaluateInst(I, Cache);
+      if (!eval.hasValue() && Root->K != souper::Inst::Select) {
+        return Result;
+        // result is `Unavailable` if args fail to evaluate
+      }
+      EvaluatedArgs.push_back(eval);
+    }
+    Result = evaluateInstRecursive(Root, EvaluatedArgs);
+  }
+  return Result;
+}
+
+EvalValue getValue(Inst *I, ValueCache &C) {
+  if (I->K == souper::Inst::Const) {
+      return {I->Val};
+  } else if (I->K == souper::Inst::Var
+      || I->K == souper::Inst::ReservedConst
+      || I->K == souper::Inst::ReservedInst) {
+    if (I->Name != "" && C.find(I) != C.end()) {
+      return C[I];
+    } else {
+      return EvalValue(); // unavailable
+    }
+  }
+  return EvalValue();
+}
+
+llvm::KnownBits dataflow::findKnownBits(Inst* I, ValueCache& C) {
+  llvm::KnownBits Result(I->Width);
+  switch(I->K) {
+    case souper::Inst::Const:
+    case souper::Inst::Var : {
+      EvalValue V = getValue(I, C);
+      if (V.hasValue()) {
+        Result.One = V.getValue();
+        Result.Zero = ~V.getValue();
+        return Result;
+      } else {
+        return Result;
+      }
+    }
+    case souper::Inst::Shl : {
+      auto Op0KB = findKnownBits(I->Ops[0], C);
+      auto Op1V = getValue(I->Ops[1], C);
+      if (Op1V.hasValue()) {
+        Op0KB.One <<= Op1V.getValue();
+        Op0KB.Zero <<= Op1V.getValue();
+        Op0KB.Zero.setLowBits(Op1V.getValue().getLimitedValue());
```

Comment:
what's the rationale for getLimitedValue being correct here? the shift exponent can't be that big I guess? 

I think any use of getLimitedValue in bitwidth-portable code deserves a comment explaining why it's correct